### ISSUE TITLE
Confusion Matrix

### ISSUE THREAD

**Issue by Muaazbinsaeed** [tensorflow/tensorflow#1606]
Is there any operation which generates Confusion Matrix in Tensorflow?
Can't find it on http://stackoverflow.com/questions/tagged/tensorflow

**Comment by fayeshine** [comment]
You can do it in this way, maybe a little tedious

res = tf.pack([tf.argmax(y,1), tf.argmax(y_,1)])
ans = res.eval(feed_dict={x: mnist.test.images, y_: mnist.test.labels})
import numpy as np
import pandas as pd
confusion = np.zeros([10,10],int)
for p in ans.T:
    confusion[p[0],p[1]]+=1
print(pd.DataFrame(confusion))

**Comment by Muaazbinsaeed** [comment]
@fayeshine thanks
There could be many ways to implement it using other libraries 
but that's not the way tensorflow is.
Tensorflow should make this operation too.

**Comment by girving** [comment]
Fixed by #1999.

**Comment by syedk01** [comment]
I am hoping to print the confusion matrix in my code
In the line where it says 
Tgtc=11
How do i add something like 
Tgtc=[2,5,7,8,11] where i want to do a confusion matrix for the 5 variables

>>>>>>>>>>>>
import sys

if sys.version_info.major != 3:
    raise Exception('Please run with python3')

import os
import time
import gc
import numpy as np

import afib_common

from afib_common import (
    build_model, get_batch, train_test_split, get_batch2)
from afib_data import reload_settings

# Load constants defined in afib_data
reload_settings()
if not os.path.isdir('modelsave'):
    os.path.mkdir('modelsave')
model = build_model()

def excld(N, i):
    ''' Generates a list of numbers 0..N-1 without the number i'''
    return tuple(_ for _ in range(N) if _ != i)

# Target class for performance metric
tgtC = 11
# (normal = "N", is the 11th item in afib_common.symbols_keep)

def calc_confusion_matrix(test_y, test_yhat):
    ''' Generates a confusion matrix from actual classes and predicted classes.

    Columns are the predicted class, while rows are the actual.
    '''
    N = test_y.shape[1]
    cm = np.zeros((N, N))
    cls_y = test_y.argmax(1)
    cls_yhat = test_yhat.argmax(1)
    for actual, predicted in zip(cls_y, cls_yhat):
        cm[actual, predicted] += 1
    return cm / cm.sum()

def estimate_f1_score():
    ''' Estimate the F1 score using many samples.

    Numbrer of samples is controlled by the Nbatch constant here.
    The F1 score is evaluated for a specific class. (See tgtC variable defined
    earlier in this file.)

    F1 definition: https://en.wikipedia.org/wiki/F1_score

    PS: This function uses the class prevalence as returned by `get_batch`.
    (I.e. all classes are sampled with probability 1/NC where NC = number of
    classes)
    '''
    Nbatch = 1000
    test_batch = get_batch(test_records, Nbatch)
    test_yhat = model.predict(test_batch.X)
    true_positive = (
        (test_batch.y.argmax(1) == tgtC) *
        (test_yhat.argmax(1) == tgtC)).sum()
    false_positive = (
        (test_batch.y.argmax(1) != tgtC) *
        (test_yhat.argmax(1) == tgtC)).sum()
    false_negative = (
        (test_batch.y.argmax(1) == tgtC) *
        (test_yhat.argmax(1) != tgtC)).sum()
    # true_negative = (
    #     (test_batch.y.argmax(1) != tgtC) *
    #     (test_yhat.argmax(1) != tgtC)).sum()
    f1 = 2 * true_positive / (
        2 * true_positive + false_positive + false_negative)
    conf_matrix = calc_confusion_matrix(test_batch.y, test_yhat)
    normal_acc = calc_normal_accuracy(conf_matrix)
    return f1, normal_acc, conf_matrix

# NC = Number of classes
NC = len(afib_common.symbols_keep)

def generate_proper_test_train_split():
    '''Prepare a train-test split such that the split is across patients.

    This prevents that a data leak occurs where the model learns to identify
    patients instead of heartbeat classes.

    Although, this makes it more difficult to find a split where the amount
    of beats in the training set vs testing set is not close to 80%:20% that
    we would otherwise easily get with a naive split.

    The approach here is to randomly repeat the split until the beats are
    split close enough to 80:20.
    (For some classes this is impossible, and we settle for anything between
    40% to 95% in the training set)
    '''
    problems = 1
    retries = -1
    while problems != 0:
        retries += 1
        print(
            'Generating properly balanced test-train split. Try#', retries + 1)
        train_records, test_records = train_test_split()
        problems = 0
        prob_fracs = []
        for k in train_records:
            len_train = len(train_records[k])
            len_test = len(test_records[k])
            frac = len_train / (len_train + len_test)
            if frac < 0.4 or frac > 0.95:  # want 0.8 mostly...
                problems += 1
                prob_fracs.append(frac)
            # print(retries, problems, prob_fracs)
    VIEW_SPLIT_STATS = 0
    if VIEW_SPLIT_STATS:
        for k in train_records:
            len_train = len(train_records[k])
            len_test = len(test_records[k])
            frac = len_train / (len_train + len_test)
            print(k, '   ', '%3f' % frac, '   ', len_train + len_test)
    return train_records, test_records

def conf_matrix_for_class(conf_matrix, tgtC):
    ''' Get the 2x2 confidence matrix for a specific class given the confidence
        matrix for all the classes.
    '''
    tp = conf_matrix[tgtC, tgtC]
    tn = conf_matrix[np.ix_(excld(NC,tgtC),excld(NC,tgtC))].sum()
    fp = conf_matrix[np.ix_(excld(NC, tgtC),[tgtC])].sum()
    fn = conf_matrix[np.ix_([tgtC], excld(NC, tgtC))].sum()
    return np.array([[tp, fn], [fp, tn]])

def adjust_prevalence(conf_matrix_2x2, prevl_new):
    ''' Takes a 2x2 confidence matrix and adjusts the rows so that the
        prevalnce of the positive class becomes `prevl_new`

        Prevalence = (TP + FN) / (TP + FN + FP + TN)
    '''
    tmp = conf_matrix_2x2
    tmp = tmp / tmp.sum(1, keepdims=1)
    tmp = tmp * [[prevl_new], [1 - prevl_new]]
    tmp = tmp / tmp.sum()
    return tmp

def conf_matrix_metrics(conf_mat_2x2):
    ''' Takes a 2x2 confidence matrix and returns the various binary metrics
        associated with it as a dictionary.

    definitions: https://en.wikipedia.org/wiki/Sensitivity_and_specificity
    '''
    (tp, fn), (fp, tn) = conf_mat_2x2
    acc = tp + tn
    sens = tp / (tp + fn)
    prec = tp / (tp + fp)
    prev = tp + fn
    f1 = 2 * tp / (2 * tp + fn + fp)
    return {'tp': tp, 'tn': tn, 'fp': fp, 'fn': fn,
            'accuracy': acc, 'sensitivity': sens, 'precision': prec,
            'prevalence': prev, 'f1': f1}

def calc_normal_accuracy(conf_matrix):
    ''' Takes a confidence matrix, converts it to the 2x2 specific confidence
        matrix of class `tgtC`(normal) and rescales the prevalence according
        to the prevalence of the normal class in the databases.
        (At the time of writing 96%)
    '''
    normal_prevalence = 0.96  # Prevalence of "normal" beats in the datasets
    tmp0 = conf_matrix_for_class(conf_matrix, tgtC)
    tmp1 = adjust_prevalence(tmp0, normal_prevalence)
    return conf_matrix_metrics(tmp1)['accuracy']

train_records, test_records = generate_proper_test_train_split()

score_hist = []
score_hist2f1 = []  # predictive f1
score_hist2acc = []  # predictive accuracy
IT1 = 20000  # Number of batches to train on before exiting

class_acc_hist = []  # Per-class accuracies
dt_hist = []  # timing performance history (for debugging)

print('\nStarting training...')
k = 0
while len(score_hist) < IT1:
    t1 = time.time()
    batch = get_batch2(train_records, 16)
    batch_y_soft = batch.y * 0.99 + 0.01 / batch.y.shape[1]  # soften labels
    score = model.train_on_batch(batch.X, batch_y_soft)
    score_hist.append(score[0])
    t2 = time.time()
    dt_hist.append(t2 - t1)
    if len(score_hist) % 50 == 1:
        gc.collect()  # Helps - although it shouldn't :(
        f1_pred, acc_pred, conf_matrix = estimate_f1_score()
        class_acc_hist.append(
            conf_matrix.diagonal() / conf_matrix.sum(1))  # sensitivity
        print('Batches seen:', len(score_hist))
        print('    f1, acc = %r' % ((f1_pred, acc_pred),))
        score_hist2f1.append(f1_pred)
        score_hist2acc.append(acc_pred)

print('Saving trained model...')
model.save('modelsave/model.data')
print('Done!')
